util/ranger/ranger.go

```diff
@@ -88,9 +90,12 @@ func points2Ranges(sctx sessionctx.Context, rangePoints []*point, tp *types.Fiel
 			HighExclude: endPoint.excl,
 			Collators:   []collate.Collator{collate.GetCollator(tp.GetCollate())},
 		}
+		if len(ranges) == 0 && rangeMaxSize > 0 && ran.MemUsage()*int64(len(rangePoints))/2 > rangeMaxSize {
+			return FullRange(), true, nil
```

Comment:
I think we can record fallback flag in `points2Ranges` instead of return `rangeFallback` here

Reply:
If we don't return `rangeFallback`, it is not convenient for the caller of `points2Ranges`(such as `buildColumnRange`) to check whether range fallback happens in `points2Ranges`. If range fallback happens, the caller can early return, otherwise it needs to union ranges before return. We cannot use the flag `StatementContext.RangeFallback` to check whether range fallback happens because the flag may be set by another building range procedure in optimization of the same query.